# In this file, we use Multiple Linear Regression, prior to the feature reduction process to test how well the model can predict the target.  This does not include every column from the raw datasets but will, nevertheless, include every column that remains following the basic preliminary cleaning stage.  Thus, we will import and merge each of the 3 cleaned data tables from postgres to perform this "baseline" testing procedure.

In [16]:
import pandas as pd
import numpy as np
from sqlalchemy import create_engine
import psycopg2
import time

In [36]:
from pathlib import Path
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn import linear_model
from sklearn.model_selection import train_test_split
from sklearn.metrics import balanced_accuracy_score

In [18]:
# Download raw data from postgres for stage 1 ETL

conn_string = 'postgres://whnpmxwsiccrtg:53c453893549d2b1e6a4ff92e626a2a08ebcaff66678e50d33e3742f66e3e4f4@ec2-52-4-171-132.compute-1.amazonaws.com/d2ajro4cjr10lb'

db = create_engine(conn_string)
conn = db.connect()

start_time = time.time()
clean_listing = pd.read_sql_query('select * from "listings_full_clean"',con=conn)
calendar = pd.read_sql_query('select * from "calendar_clean"',con=conn)
amenities = pd.read_sql_query('select * from "amenities_clean"',con=conn)
print("PostGres Download Duration: {} seconds".format(time.time() - start_time))


PostGres Download Duration: 1.8125574588775635 seconds


In [19]:
# View clean_listing database
clean_listing.head()

id last_scraped host_since  host_listings_count host_is_superhost  \
0   6422   2019-01-17 2009-04-03                    1                 t   
1  20847   2019-01-17 2010-02-08                    2                 t   
2  25341   2019-01-17 2010-02-08                    2                 t   
3  25613   2019-01-17 2010-03-18                    1                 t   
4  37306   2019-01-17 2010-07-08                    1                 f   

  host_identity_verified neighbourhood_cleansed   latitude  longitude  \
0                      t             District 6  36.173151 -86.735813   
1                      f             District 7  36.196397 -86.704267   
2                      f             District 7  36.197053 -86.704937   
3                      t             District 6  36.178263 -86.741622   
4                      f            District 17  36.133109 -86.774767   

      room_type  ... review_scores_checkin  review_scores_value  \
0  Private room  ...                  10.0                 10.0   
1  Private room  ...                   9.0                  9.0   
2  Private room  ...                  10.0                  9.0   
3  Private room  ...                  10.0                 10.0   
4  Private room  ...                  10.0                 10.0   

   instant_bookable  is_business_travel_ready          cancellation_policy  \
0                 f                         f  strict_14_with_grace_period   
1                 f                         f  strict_14_with_grace_period   
2                 f                         f  strict_14_with_grace_period   
3                 t                         f                     moderate   
4                 f                         f  strict_14_with_grace_period   

   require_guest_profile_picture  require_guest_phone_verification  \
0                              f                                 f   
1                              t                                 t   
2                              t                                 t   
3                              t                                 t   
4                              t                                 t   

   has_availability  is_review  days_host  
0                 t       True       3576  
1                 t       True       3265  
2                 t       True       3265  
3                 t       True       3227  
4                 t       True       3115  

[5 rows x 33 columns]

In [20]:
# View data types of full_listings
clean_listing.dtypes

id                                           int64
last_scraped                        datetime64[ns]
host_since                          datetime64[ns]
host_listings_count                          int64
host_is_superhost                           object
host_identity_verified                      object
neighbourhood_cleansed                      object
latitude                                   float64
longitude                                  float64
room_type                                   object
property_type                               object
accommodates                                 int64
bathrooms                                  float64
bedrooms                                     int64
bed_type                                    object
price                                      float64
security_deposit                           float64
cleaning_fee                               float64
review_scores_rating                       float64
number_of_reviews              

In [21]:
# Drop duplicated variables.
listing = clean_listing.drop(columns = ['last_scraped', 'host_since', 'latitude', 'longitude', 'price'])

In [22]:
# View amenities dataframe
amenities.head()

id   tv  internet  wifi  air_conditioning  kitchen  \
0   6422  1.0       1.0   1.0               1.0      1.0   
1  20847  1.0       0.0   1.0               1.0      1.0   
2  25341  1.0       0.0   1.0               1.0      1.0   
3  25613  1.0       1.0   1.0               1.0      1.0   
4  37306  1.0       0.0   1.0               1.0      1.0   

   free_parking_on_premises  pets_live_on_this_property  dogs  cats  ...  \
0                       1.0                         1.0   1.0   1.0  ...   
1                       1.0                         0.0   0.0   0.0  ...   
2                       1.0                         0.0   0.0   0.0  ...   
3                       1.0                         0.0   0.0   0.0  ...   
4                       0.0                         1.0   1.0   1.0  ...   

   beachfront  ski-in/ski-out  beach_view  high-resolution_computer_monitor  \
0         0.0             0.0         0.0                               0.0   
1         0.0             0.0         0.0                               0.0   
2         0.0             0.0         0.0                               0.0   
3         0.0             0.0         0.0                               0.0   
4         0.0             0.0         0.0                               0.0   

   standing_valet  pool_with_pool_hoist  sun_loungers  electric_profiling_bed  \
0             0.0                   0.0           0.0                     0.0   
1             0.0                   0.0           0.0                     0.0   
2             0.0                   0.0           0.0                     0.0   
3             0.0                   0.0           0.0                     0.0   
4             0.0                   0.0           0.0                     0.0   

   hammock  mobile_hoist  
0      0.0           0.0  
1      0.0           0.0  
2      0.0           0.0  
3      0.0           0.0  
4      0.0           0.0  

[5 rows x 173 columns]

In [23]:
# View calendar dataframe
calendar.head(10)

id      day  month  price
0  6422  weekday      1   40.0
1  6422  weekday      2   40.0
2  6422  weekday      3   40.0
3  6422  weekday      4   40.0
4  6422  weekday      5   40.0
5  6422  weekday      6   40.0
6  6422  weekday      7   40.0
7  6422  weekday      8   40.0
8  6422  weekday      9   40.0
9  6422  weekday     10   40.0

In [24]:
# Merge amenities table to full listings.
merge1 = listing.merge(amenities, how='left', on ='id')

In [25]:
# Merge calendar table to amenities and full listings.
merged = merge1.merge(calendar, how='left', on ='id')
merged.head(20)

id  host_listings_count host_is_superhost host_identity_verified  \
0   6422                    1                 t                      t   
1   6422                    1                 t                      t   
2   6422                    1                 t                      t   
3   6422                    1                 t                      t   
4   6422                    1                 t                      t   
5   6422                    1                 t                      t   
6   6422                    1                 t                      t   
7   6422                    1                 t                      t   
8   6422                    1                 t                      t   
9   6422                    1                 t                      t   
10  6422                    1                 t                      t   
11  6422                    1                 t                      t   
12  6422                    1                 t                      t   
13  6422                    1                 t                      t   
14  6422                    1                 t                      t   
15  6422                    1                 t                      t   
16  6422                    1                 t                      t   
17  6422                    1                 t                      t   
18  6422                    1                 t                      t   
19  6422                    1                 t                      t   

   neighbourhood_cleansed     room_type property_type  accommodates  \
0              District 6  Private room         House             2   
1              District 6  Private room         House             2   
2              District 6  Private room         House             2   
3              District 6  Private room         House             2   
4              District 6  Private room         House             2   
5              District 6  Private room         House             2   
6              District 6  Private room         House             2   
7              District 6  Private room         House             2   
8              District 6  Private room         House             2   
9              District 6  Private room         House             2   
10             District 6  Private room         House             2   
11             District 6  Private room         House             2   
12             District 6  Private room         House             2   
13             District 6  Private room         House             2   
14             District 6  Private room         House             2   
15             District 6  Private room         House             2   
16             District 6  Private room         House             2   
17             District 6  Private room         House             2   
18             District 6  Private room         House             2   
19             District 6  Private room         House             2   

    bathrooms  bedrooms  ... high-resolution_computer_monitor  standing_valet  \
0         1.0         2  ...                              0.0             0.0   
1         1.0         2  ...                              0.0             0.0   
2         1.0         2  ...                              0.0             0.0   
3         1.0         2  ...                              0.0             0.0   
4         1.0         2  ...                              0.0             0.0   
5         1.0         2  ...                              0.0             0.0   
6         1.0         2  ...                              0.0             0.0   
7         1.0         2  ...                              0.0             0.0   
8         1.0         2  ...                              0.0             0.0   
9         1.0         2  ...                              0.0             0.0   
10        1.0         2  ...                              0.0       

In [26]:
# Merge tables and drop 'id' (unique identifier - not relevant)
merged = merged.drop(columns = 'id')

In [27]:
# Create a copy of the database to work with going forward.
merged2 = merged.copy()

In [28]:
# Create a list of the object type columns.
objects = merged2.dtypes[merged2.dtypes == 'object'].index.tolist()
objects

['host_is_superhost',
 'host_identity_verified',
 'neighbourhood_cleansed',
 'room_type',
 'property_type',
 'bed_type',
 'instant_bookable',
 'is_business_travel_ready',
 'cancellation_policy',
 'require_guest_profile_picture',
 'require_guest_phone_verification',
 'has_availability',
 'day']

In [29]:
# Encode objects to transform objects type columns to numeric values.
le = LabelEncoder()
for obj in objects:
    merged2[obj] = le.fit_transform(merged2[obj])

In [30]:
# View the merged dataframe
merged2.head()

host_listings_count  host_is_superhost  host_identity_verified  \
0                    1                  1                       1   
1                    1                  1                       1   
2                    1                  1                       1   
3                    1                  1                       1   
4                    1                  1                       1   

   neighbourhood_cleansed  room_type  property_type  accommodates  bathrooms  \
0                      31          1             17             2        1.0   
1                      31          1             17             2        1.0   
2                      31          1             17             2        1.0   
3                      31          1             17             2        1.0   
4                      31          1             17             2        1.0   

   bedrooms  bed_type  ...  high-resolution_computer_monitor  standing_valet  \
0         2         4  ...                               0.0             0.0   
1         2         4  ...                               0.0             0.0   
2         2         4  ...                               0.0             0.0   
3         2         4  ...                               0.0             0.0   
4         2         4  ...                               0.0             0.0   

   pool_with_pool_hoist  sun_loungers  electric_profiling_bed  hammock  \
0                   0.0           0.0                     0.0      0.0   
1                   0.0           0.0                     0.0      0.0   
2                   0.0           0.0                     0.0      0.0   
3                   0.0           0.0                     0.0      0.0   
4                   0.0           0.0                     0.0      0.0   

   mobile_hoist  day  month  price  
0           0.0    0      1   40.0  
1           0.0    0      2   40.0  
2           0.0    0      3   40.0  
3           0.0    0      4   40.0  
4           0.0    0      5   40.0  

[5 rows x 202 columns]

In [31]:
pd.pandas.set_option('display.max_columns', None)

In [32]:
merged2.head()

561                       10.0                    10.0   
2                561                       10.0                    10.0   
3                561                       10.0                    10.0   
4                561                       10.0                    10.0   

   review_scores_communication  review_scores_checkin  review_scores_value  \
0                         10.0                   10.0                 10.0   
1                         10.0                   10.0                 10.0   
2                         10.0                   10.0                 10.0   
3                         10.0                   10.0                 10.0   
4                         10.0                   10.0                 10.0   

   instant_bookable  is_business_travel_ready  cancellation_policy  \
0                 0                         0                    3   
1                 0                         0                    3   
2                 0                         0                    3   
3                 0                         0                    3   
4                 0                         0                    3   

   require_guest_profile_picture  require_guest_phone_verification  \
0                              0                                 0   
1                              0                                 0   
2                              0                                 0   
3                              0                                 0   
4                              0                                 0   

   has_availability  is_review  days_host   tv  internet  wifi  \
0                 0       True       3576  1.0       1.0   1.0   
1                 0       True       3576  1.0       1.0   1.0   
2                 0       True       3576  1.0       1.0   1.0   
3                 0       True       3576  1.0       1.0   1.0   
4                 0       True       3576  1.0       1.0   1.0   

   air_conditioning  kitchen  free_parking_on_premises  \
0               1.0      1.0                       1.0   
1               1.0      1.0                       1.0   
2               1.0      1.0                       1.0   
3               1.0      1.0                       1.0   
4               1.0      1.0                       1.0   

   pets_live_on_this_property  dogs  cats  heating  family/kid_friendly  \
0                         1.0   1.0   1.0      1.0                  1.0   
1                         1.0   1.0   1.0      1.0                  1.0   
2                         1.0   1.0   1.0      1.0                  1.0   
3                         1.0   1.0   1.0      1.0                  1.0   
4                         1.0   1.0   1.0      1.0                  1.0   

   washer  dryer  smoke_detector  carbon_monoxide_detector  first_aid_kit  \
0     1.0    1.0             1.0                       1.0            1.0   
1     1.0    1.0             1.0                       1.0            1.0   
2     1.0    1.0             1.0                       1.0            1.0   
3     1.0    1.0             1.0                       1.0            1.0   
4     1.0    1.0             1.0                       1.0            1.0   

   safety_card  fire_extinguisher  essentials  shampoo  lock_on_bedroom_door  \
0          1.0                1.0         1.0      1.0                   1.0   
1          1.0                1.0         1.0      1.0                   1.0   
2          1.0                1.0         1.0      1.0                   1.0   
3          1.0                1.0         1.0      1.0                   1.0   
4          1.0                1.0         1.0      1.0                   1.0   

   hangers  hair_dryer  iron  laptop_friendly_workspace  bathtub  \
0      1.0         1.0   1.0                        1.0      1.0   
1      1.0         1.0   1.0                        1.0      1.0   
2      1.0         1.0   1.0                        1.0     

In [33]:
# Define the target variable.
y=merged2['price']

In [34]:
# Define the features.
X = merged2.drop(columns ='price')
X = X.values.reshape(-1, 201)

In [37]:
# Scale the feature data.
data_scaler = StandardScaler()
scaled_X = data_scaler.fit_transform(X)

In [38]:
# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(scaled_X, y, random_state=78)

In [39]:
# Fit the linear regression model to the data.
regr = linear_model.LinearRegression()
regr.fit(scaled_X, y)

LinearRegression()

In [40]:
# Calculate the predicted prices for scaled_X using the model.
y_pred = regr.predict(scaled_X)
print(y_pred)

[-75.10879827 -71.07153924 -67.03428021 ... 580.31310803 584.34858633
 588.38584536]


In [41]:
# Calculate the accuracy score of the initial model.
regr.score(scaled_X, y, sample_weight=None)


0.5719488379871291

# Multiple linear regression was only able to predict the target with about 57% accuracy